---
title: "Wrangling data for input into `MatrixEQTL`"
author: "Aine Fairbrother-Browne"
date: "11/21"
output: 
  html_document:
    theme: paper
    highlight: kate
    df_print: paged
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: false  
    fig_width: 20
    fig_height: 15
    fig_caption: true
---


**Aim**: To wrangle covariate, transcriptomic and methylation data into the correct format for input into `MatrixEQTL`   

# Setup  

```{r setup, include=FALSE}

# import libs
library(tidyverse)
library(vroom)
library(EnsDb.Hsapiens.v79)

# notebook settings 
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/home/abrowne/projects/amppd_analysis/")

```

# Covariates  

* This input to `MatrixEQTL` will consist of a single file containing sample metadata, PEER factors and gPCs   
* The format will be as such: rows=variables, cols=samples  

## Metadata    

* The function `wrangle.filter.export.meta()` does the following:  
+ Filter for cohort  
+ Filter for diagnosis  
+ Filter for release `1015` samples  
+ Select only the `sample_id`, `age` and `sex` variables  
+ Encode `sex`: Female ~ 0, Male ~ 1  
+ Transpose --> rows=meta, cols=samples   

```{r}

meta = readr::read_csv("~/projects/amppd_analysis/metadata/2020_v2release_1218/filtered/amppd-meta-cc-demo-rnaseq-medhist-variable-missing-n=25-filt-ALL-timepoints.csv", show_col_types = FALSE)

meta %>% 
  dplyr::group_by(cohort, visit_month) %>% 
  dplyr::summarise(n=n())

```

```{r}

wrangle.filter.export.meta = function(cohort, diag){
  
  print(paste(cohort, diag))
  
  cohort="PP"
  diag="Case"
  
  if(diag != "cohort"){
    
    samples.file = read_csv(paste0("./data/AMPPD_sample_lists/",cohort,"-visit_month=ALL-case_control_other_latest==",diag,"_unique_1015.csv"), show_col_types = FALSE) %>% 
      dplyr::pull(x)
    
    meta = readr::read_csv("./metadata/2020_v2release_1218/filtered/amppd-meta-cc-demo-rnaseq-medhist-variable-missing-n=25-filt-ALL-timepoints.csv", show_col_types = FALSE) %>% 
      dplyr::filter(cohort==cohort) %>% 
      dplyr::filter(case_control_other_latest==diag) %>% 
      dplyr::select(sample_id, sex, age_at_baseline) %>% 
      dplyr::filter(sample_id %in% samples.file) %>% 
      dplyr::distinct() %>% 
      dplyr::mutate(sex=case_when(
        sex=="Male" ~ 0,
        sex=="Female" ~ 1
      )) %>% 
      tibble::column_to_rownames("sample_id") %>% 
      t() %>% 
      as.data.frame() %>% 
      tibble::rownames_to_column("variable")
    
    meta %>% dim() %>% print()
    
    meta %>% 
      readr::write_csv(x=., file=paste0("./data/MatrixEQTL_input/covariates/all_timepoints/",cohort,"_",diag,"_timepoint_averaged_meta_age_sex_for_MatrixEQTL.csv"))
  }
  
  if(diag == "cohort"){
    
    samples.file = read_csv(paste0("./data/AMPPD_sample_lists/",cohort,"-visit_month=ALL-case_control_other_latest!=Other_unique_1015.csv"), show_col_types = FALSE) %>% 
      dplyr::pull(x)
    
    meta = readr::read_csv("./metadata/2020_v2release_1218/filtered/amppd-meta-cc-demo-rnaseq-medhist-variable-missing-n=25-filt-ALL-timepoints.csv", show_col_types = FALSE) %>% 
      dplyr::filter(cohort==cohort) %>% 
      dplyr::select(sample_id, sex, age_at_baseline, case_control_other_latest) %>% 
      dplyr::filter(sample_id %in% samples.file) %>% 
      dplyr::distinct() %>% 
      dplyr::mutate(
        
        sex=case_when(
          sex=="Male" ~ 0,
          sex=="Female" ~ 1
        ),
        case_control_other_latest=case_when(
          case_control_other_latest=="Case" ~ 1,
          case_control_other_latest=="Control" ~ 0 
        )
        
      ) %>% 
      tibble::column_to_rownames("sample_id") %>% 
      t() %>% 
      as.data.frame() %>% 
      tibble::rownames_to_column("variable")
    
    meta %>% dim() %>% print()
    
    meta %>% 
      readr::write_csv(x=., file=paste0("./data/MatrixEQTL_input/covariates/all_timepoints/",cohort,"_",diag,"_timepoint_averaged_meta_age_sex_for_MatrixEQTL.csv"))
  }
}

```

```{r message=FALSE, warning=FALSE}

for(cohort in c("PP","PD")){
  for(diag in c("Case", "Control", "cohort")){
    wrangle.filter.export.meta(cohort=cohort, diag=diag)
  }
}

```

## PEER factors  

* The function `wrangle.filter.export.peer.matrices()` does the following:  
+ Import  
+ Filter for release `1015` samples  
+ Re-label PEER factors so it's clear what they are in the later merged cov. file  
+ Select only the first 10 PEER factors  
+ Transpose --> rows=factors, cols=samples   

```{r message=FALSE, warning=FALSE}

wrangle.filter.export.peer.matrices = function(fname){
  
  basename=gsub("_timepoint_averaged_PEER_wrangled.csv", "", fname)
  cohort=stringr::str_extract(basename, "^.{2}")
  diag=gsub(cohort, "", basename) %>% gsub("_", "", .)
  
  print(paste(cohort, diag))
  
  # read in relevant sample_id list
  if(diag == "cohort"){
    new.diag="cohort"
    samples.file = read_csv(paste0("./data/AMPPD_sample_lists/",cohort,"-visit_month=ALL-case_control_other_latest!=Other_unique_1015.csv"), show_col_types = FALSE) %>% 
      dplyr::pull(x)
  }
  if(diag == "case"){
    new.diag="Case"
    samples.file = read_csv(paste0("./data/AMPPD_sample_lists/",cohort,"-visit_month=ALL-case_control_other_latest==Case_unique_1015.csv"), show_col_types = FALSE) %>% 
      dplyr::pull(x)
  }
  if(diag == "ctrl"){
    new.diag="Control"
    samples.file = read_csv(paste0("./data/AMPPD_sample_lists/",cohort,"-visit_month=ALL-case_control_other_latest==Control_unique_1015.csv"), show_col_types = FALSE) %>% 
      dplyr::pull(x)
  }
  
  # read in relevant peer file
  peer.file = readr::read_csv(paste0("./data/PEER_output/rosalind-cmd-line-gen/",fname), show_col_types = FALSE) %>% 
    tibble::column_to_rownames("...1") %>%
    `colnames<-`(gsub("V", "PEER", colnames(.))) %>% # relabel so it's clear they're PEER factors
    dplyr::select(1:10) %>% # get first 10 PEER factors
    t(.) %>% # transpose
    as.data.frame(.) %>% 
    tibble::rownames_to_column("variable")
  
  # peer.file %>% dim() %>% print()
  # peer.file %>% .[1:5,1:5] %>% print()
  
  peer.file %>% 
    readr::write_csv(x=., file=paste0("./data/MatrixEQTL_input/covariates/all_timepoints/",cohort,"_",new.diag,"_10_timepoint_averaged_PEER_for_MatrixEQTL.csv"))
  
}

```

```{r message=FALSE, warning=FALSE}

for(i in list.files(path="./data/PEER_output/rosalind-cmd-line-gen", pattern="timepoint_averaged_PEER_wrangled.csv", full.names=F)){
  wrangle.filter.export.peer.matrices(fname=i)
}

```

* Check sizes of output meta files  

```{r}

for(i in list.files(path="./data/MatrixEQTL_input/covariates/all_timepoints/", pattern="_10_timepoint_averaged_PEER_for_MatrixEQTL.csv", full.names=T)){
  print(i)
  vroom::vroom(file=i, show_col_types = FALSE) %>% dim() %>% print()
}

```

## gPCs  

```{r}

generate.matrixeqtl.gpc.file = function(cohort, diag){
  
  if(diag=="Case"){sample.id.file.label="latest==Case"}
  if(diag=="Control"){sample.id.file.label="latest==Control"}
  if(diag=="cohort"){sample.id.file.label="latest!=Other"}
  
  data.out="./data/bigsnpr_gPC_output/all_timepoints/"
  
  samples.to.use = readr::read_csv(paste0("./data/AMPPD_sample_lists/",cohort,"-visit_month=ALL-case_control_other_",sample.id.file.label,"_unique_1015.csv"), show_col_types = FALSE) %>% 
    dplyr::pull(x) %>% 
    stringr::str_extract(., "P.+-") %>% 
    gsub("-$","",.) %>% 
    unique()
  
  # Get sample IDs
  sample.ids = obj.bigSNP$fam %>% 
    dplyr::pull(sample.ID)
  
  # Add sample ids to sample-PC df and write out
  svd0$u %>%
    as.data.frame() %>% 
    `rownames<-`(sample.ids) %>%
    .[rownames(.) %in% samples.to.use, 1:5] %>% 
    rownames_to_column("sample_id") %>% 
    readr::write_csv(x=., file=paste0(data.out, cohort, "_", diag, "_svd.obj$u_5_gPCs_maf0.05_1015.csv"))
  
}

```

```{r}

for(cohort in c("PP", "PD")){
  
  load(paste0("./data/bigsnpr_gPC_output/",cohort,"_bigsnpr_data.RData"))
  
  for(diag in c("Case", "Control", "cohort")){
    generate.matrixeqtl.gpc.file(cohort=cohort, diag=diag)
  }
}

```

* Make file into the right shape for `MatrixEQTL`  

```{r}

wrangle.filter.export.gpcs = function(cohort, diag){
  
  print(paste(cohort, diag))
  
  gpc.file = read.csv(paste0("./data/bigsnpr_gPC_output/all_timepoints/",cohort,"_",diag,"_svd.obj$u_5_gPCs_maf0.05_1015.csv")) %>% 
    tibble::column_to_rownames("sample_id") %>% 
    `colnames<-`(gsub("V", "gPC", colnames(.))) %>% # relabel so it's clear they're PEER factors
    as.data.frame(.) 
  
  gpc.file.diag = gpc.file %>% 
    dplyr::select(gPC1, gPC2, gPC3, gPC4, gPC5) %>%  # select first 5 gPCs only
    t() %>% 
    as.data.frame() %>% 
    tibble::rownames_to_column("variable")
  
  gpc.file.diag %>% 
    dim() %>% 
    print()
  
  gpc.file.diag %>% 
    readr::write_csv(x=., file=paste0("./data/MatrixEQTL_input/covariates/all_timepoints/",cohort,"_",diag,"_5_gPCs_maf0.05_for_MatrixEQTL.csv"))
  
}

```

```{r}

for(cohort in c("PP", "PD")){
  for(diag in c("Case", "Control", "cohort")){
    wrangle.filter.export.gpcs(cohort=cohort, diag=diag)
  }
}

```

## Collect and combine covariate files (meta, PEER, gPC) into single covariate file  

* Need to generate 8 files, containing gPCs, metadata and PEER factors  
* For: PD-case, PD-control, PP-case, PP-control, PD-cohort, PP-cohort  
* For this, need to get the set of sample which has genotyping data, has age and sex metadata and has trasncriptomic data (PEER carried out using TPM matrices)  
* Then produce the final covariate table  

```{r}

generate.covariate.file = function(cohort, diag, data.type){
  
  print(paste(cohort, diag))
  
  gpc.file = readr::read_csv(paste0("./data/MatrixEQTL_input/covariates/all_timepoints/",cohort,"_",diag,"_5_gPCs_maf0.05_for_MatrixEQTL.csv"), show_col_types = FALSE) %>% 
    tibble::column_to_rownames("variable")
  
  peer.file = readr::read_csv(paste0("./data/MatrixEQTL_input/covariates/all_timepoints/",cohort,"_",diag,"_10_timepoint_averaged_PEER_for_MatrixEQTL.csv"), show_col_types = FALSE) %>% 
    tibble::column_to_rownames("variable")
  
  meta.file = readr::read_csv(paste0("./data/MatrixEQTL_input/covariates/all_timepoints/",cohort,"_",diag,"_timepoint_averaged_meta_age_sex_for_MatrixEQTL.csv"), show_col_types = FALSE) %>%
    tibble::column_to_rownames("variable") %>% 
    t() %>% 
    as.data.frame() %>% 
    tibble::rownames_to_column("variable") %>% 
    dplyr::mutate(participant_id = stringr::str_extract(variable, "P.+-") %>% gsub("-$","",.)) %>% 
    dplyr::select(-variable) %>% 
    distinct() %>% 
    tibble::column_to_rownames("participant_id") %>% 
    t() %>% 
    as.data.frame()
  
    # read in genomics - gen by 8b_wrangle_geno_mat_for_matrixeqtl.R
  if(diag=="cohort"){
    genomics.file = data.table::fread(paste0("/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/genomics/",cohort,"_all_chrs_geno_matrix_012_wrangled.csv")) %>% 
      colnames(.)
  } else{
    genomics.file = data.table::fread(paste0("/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/genomics/",cohort,"_",tolower(diag),"_all_chrs_geno_matrix_012_wrangled.csv")) %>% 
      colnames(.)
  }
    
  print(paste(
    "gpc n = ",
    gpc.file %>% colnames() %>% unique() %>% length(),
    "peer n = ",
    peer.file %>% colnames() %>% unique() %>% length(),
    "meta n = ",
    meta.file %>% colnames() %>% unique() %>% length(),
    "genomics n = ",
    genomics.file %>% unique() %>% length()
  ))
  
  sample.set = Reduce(intersect, list(
    gpc.file %>% colnames() %>% unique(),
    peer.file %>% colnames() %>% unique(),
    meta.file %>% colnames() %>% unique(),
    genomics.file %>% unique()
  ))

  print(paste(
    "sample set n = ",  
    sample.set %>% length()
  ))
  
  # write out sample.set
  sample.set %>% 
    as.data.frame() %>% 
    readr::write_csv(x=., file=paste0("./data/MatrixEQTL_input/covariates/all_timepoints/",cohort,"_",diag,"_averaged_timepoint_reduce_intersect_gpc_peer_meta_sampleid.csv"))
  
  gpc.file = gpc.file %>%
    .[sample.set]
  
  gpc.file %>%
    .[1:5,1:5]
  
  peer.file = peer.file %>%
    .[sample.set]
  
  peer.file %>%
    .[1:5,1:5]
  
  meta.file = meta.file %>%
    .[sample.set]
  
  meta.file %>%
    .[,1:5]
  
  # meta mist be last so that case_control_other_latest is the 'last' covariate
  all.covs = dplyr::bind_rows(gpc.file,
                              peer.file,
                              meta.file) %>%
    tibble::rownames_to_column("variable")
  
  all.covs %>%
    readr::write_csv(x=., file=paste0("./data/MatrixEQTL_input/covariates/all_timepoints/",cohort,"_",diag,"_averaged_timepoint_maf0.05_cov_table_for_MatrixEQTL.csv"))
  
  all.covs %>%
    dim() %>%
    print()
  
}
```

```{r}

for(cohort in c("PP","PD")){
  for(diag in c("Case", "Control", "cohort")){
    generate.covariate.file(cohort=cohort, diag=diag)
  }
}

```

# Transcriptomics  

* These files need to have the same sample set as the meta/genomic data  
* They need to have MT genes only, as we're looking at associating mito expression with nuclear variation  
* They also need to be transposed, cols=samples  

```{r}

# get mito gene list

# Convert from gene.symbol to ensembl.gene
mt.gene.sym = c("MT-ND1","MT-ND2","MT-ND3","MT-ND4L","MT-ND4","MT-ND5","MT-ND6","MT-CO2","MT-CO3","MT-ATP8","MT-CYB","MT-CO1","MT-ATP6","MT-RNR1","MT-RNR2")

mt.gene.ens = ensembldb::select(EnsDb.Hsapiens.v79, keys=mt.gene.sym, keytype = "SYMBOL", columns = c("SYMBOL","GENEID")) %>% 
  dplyr::pull(GENEID)

```

```{r}

wrangle.filter.export.transcriptomics = function(cohort, diag){
  
  print(paste(cohort, diag))
  
  samples.to.use = readr::read_csv(paste0("/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/covariates/all_timepoints/",cohort,"_",diag,"_averaged_timepoint_reduce_intersect_gpc_peer_meta_sampleid.csv"), show_col_types = FALSE) %>% 
    dplyr::pull('.')
  
  tpm = vroom::vroom(paste0("/home/abrowne/projects/amppd_analysis/data/processed_and_unprocessed_transcriptomic/mt_nuc_corr_pipeline_output/all_timepoints/",cohort,"_",diag,"_timepoint_averaged_log10_mediannorm_TPM_masked_outliers_q1-3iqr_q3+3iqr.csv"), show_col_types = FALSE) %>%
    # make compatible with samples to use
    dplyr::filter(sample_id %in% samples.to.use) %>% 
    # make sure all files have the sample sample order
    dplyr::arrange(match(sample_id, samples.to.use)) %>% 
    tibble::column_to_rownames("sample_id") %>% 
    t() %>% 
    as.data.frame() %>% 
    # make gene names non-version
    `rownames<-`(gsub("\\..*", "", rownames(.))) %>% 
    # filter for MT genes only
    .[mt.gene.ens,] %>% 
    tibble::rownames_to_column("sample_id") %>% 
    readr::write_csv(x=., file=paste0("/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/transcriptomics/all_timepoints/",cohort,"_",diag,"_timepoint_averaged_log10_mediannorm_TPM_masked_outliers_q1-3iqr_q3+3iqr_MT_ONLY.csv"))
  
}

```

```{r}

for(cohort in c("PP","PD")){
  for(diag in c("Case", "Control", "cohort")){
    wrangle.filter.export.transcriptomics(cohort=cohort, diag=diag)
  }
}

```

# Methylomics

```{r}

wrangle.filter.export.methylomics = function(cohort, diag){
  
  samples.to.use = readr::read_csv(paste0("./data/MatrixEQTL_input/covariates/all_timepoints/",cohort,"_",diag,"_averaged_timepoint_reduce_intersect_gpc_peer_meta_sampleid.csv"), show_col_types = FALSE) %>% 
    dplyr::pull('.')
  
  readr::read_csv(paste0("./data/methylation_data/all_timepoints/",cohort,"_",tolower(diag),"_all_timepoints_avg_methylation_matrix.csv"), show_col_types = FALSE) %>% 
    dplyr::filter(participant_id %in% samples.to.use) %>% 
    # make sure all files have the sample sample order
    dplyr::arrange(match(participant_id, samples.to.use)) %>% 
    tibble::column_to_rownames("participant_id") %>% 
    t() %>% 
    as.data.frame() %>% 
    tibble::rownames_to_column("participant_id") %>% 
    dplyr::rename(sample_id=participant_id) %>% 
    readr::write_csv(x=., file=paste0("./data/MatrixEQTL_input/methylomics/all_timepoints/",cohort,"_",diag,"_all_timepoints_avg_methylation_matrix_MatrixEQTL_input.csv"))
}

```

```{r}

for(cohort in c("PP","PD")){
  for(diag in c("Case", "Control", "cohort")){
    wrangle.filter.export.methylomics(cohort=cohort, diag=diag)
  }
}

```

# Check files  

```{r}

# check covs
vroom::vroom("./data/MatrixEQTL_input/covariates/all_timepoints/PP_Control_averaged_timepoint_maf0.05_cov_table_for_MatrixEQTL.csv", show_col_types = FALSE) %>% dplyr::select(contains("PP")) %>% dim()

# check transcriptomic
vroom::vroom("./data/MatrixEQTL_input/transcriptomics/all_timepoints/PP_Control_timepoint_averaged_log10_mediannorm_TPM_masked_outliers_q1-3iqr_q3+3iqr_MT_ONLY.csv", show_col_types = FALSE) %>% dplyr::select(contains("PP")) %>% dim()

# check methylation 
vroom::vroom("./data/MatrixEQTL_input/methylomics/all_timepoints/PP_Control_all_timepoints_avg_methylation_matrix_MatrixEQTL_input.csv", show_col_types = FALSE) %>% dplyr::select(contains("PP")) %>% dim()

data.table::fread(paste0("/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/genomics/all_timepoints/PP_control_all_chrs-maf0.05_geno_matrix_012_wrangled-8b.csv")) %>% 
  dim()

```

```{r}

# check covs
vroom::vroom("./data/MatrixEQTL_input/covariates/all_timepoints/PP_Control_averaged_timepoint_maf0.05_cov_table_for_MatrixEQTL.csv", show_col_types = FALSE) %>% dplyr::select(contains("PP")) %>% head(n=5)

# check transcriptomic
vroom::vroom("./data/MatrixEQTL_input/transcriptomics/all_timepoints/PP_Control_timepoint_averaged_log10_mediannorm_TPM_masked_outliers_q1-3iqr_q3+3iqr_MT_ONLY.csv", show_col_types = FALSE) %>% dplyr::select(contains("PP")) %>% head(n=5)

# check methylation 
vroom::vroom("./data/MatrixEQTL_input/methylomics/all_timepoints/PP_Control_all_timepoints_avg_methylation_matrix_MatrixEQTL_input.csv", show_col_types = FALSE) %>% dplyr::select(contains("PP")) %>% head(n=5)


```


# Split files into individual pheno files  

* This is to allow per-phenotype running of `MatrixEQTL`  

```{r message=FALSE, warning=FALSE}

source("/home/abrowne/projects/amppd_analysis/pipeline/prepare_indv_pheno_files_for_MatrixEQTL.R")

# implement function for transcriptomic data
prepare.indv.pheno.files.for.MatrixEQTL(file.path="/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/transcriptomics/all_timepoints",
                                        file.pattern="MT_ONLY",
                                        loc.for.pheno.name.in.outfile="log10")

# implement function for methylation data
prepare.indv.pheno.files.for.MatrixEQTL(file.path="/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/methylomics/all_timepoints/",
                                        file.pattern=".csv",
                                        loc.for.pheno.name.in.outfile="methylation")

```

# Filter all for samples in genomics file  

```{r}

cohort="PP"
diag="cohort"

base.dir = "./data/MatrixEQTL_input/"

# read in genomics - gen by 8b_wrangle_geno_mat_for_matrixeqtl.R
if(diag=="cohort"){
  filter.for.genomics.samples = data.table::fread(paste0("/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/genomics/",cohort,"_all_chrs_geno_matrix_012_wrangled.csv")) %>% 
    colnames(.)
} else{
  filter.for.genomics.samples = data.table::fread(paste0("/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/genomics/",cohort,"_",diag,"_all_chrs_geno_matrix_012_wrangled.csv")) %>% 
    colnames(.)
}

# filter transcriptomics


# filter methylomics

# filter covs 

```


# Session Info  

```{r}
sessionInfo()
```

