---
title: "AMPPD stratified pipeline: featurecount matrices > TPM > normalisation > generate PEER factors and correlate to metadata"
author: "Aine Fairbrother-Browne"
date: ""
output: 
  html_document:
    theme: paper
    highlight: kate
    df_print: paged
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: false  
    fig_width: 20
    fig_height: 15
    fig_caption: true
---

```{r setup, include=FALSE}

# import libs
library(tidyverse)
library(vroom)
library(purrr)
library(viridis)
library(cleandata)

# notebook settings 
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/home/abrowne/projects/amppd_analysis/")

```

# Read in AMP-PD featurecount matrix  

```{r, eval=FALSE}

count.mat = vroom("./data/processed_and_unprocessed_transcriptomic/AMPPD_featurecounts_dataframe.csv", show_col_types = FALSE) %>%
  tibble::column_to_rownames(var="Geneid") %>% 
  dplyr::rename_with(., ~gsub('\\.', '-', .)) %>% 
  tibble::rownames_to_column("gene_id") %>% 
  dplyr::mutate(gene_id = gsub("\\..*", "", gene_id)) %>%
  dplyr::distinct(gene_id, .keep_all = T) %>%
  tibble::column_to_rownames("gene_id")

count.mat[1:5,1:5]

```

```{r}

count.mat %>% dim()
count.mat %>% dplyr::select(1:10) %>% colnames(.)
count.mat %>% .[1:10,1:10] %>% rownames(.)

```

# Read in AMP-PD metadata  

```{r message=FALSE, warning=FALSE}

meta.cc.demo.rna = vroom("./metadata/metadata-cohort-tables/amppd-meta-cc-demo-rnaseq.csv", show_col_types = FALSE)
meta.cc.demo.rna.medhist = vroom("./metadata/2020_v2release_1218/filtered/amppd-meta-cc-demo-rnaseq-medhist-variable-missing-n=25-filt.csv", show_col_types = FALSE)

```

* Check sample numbers under the key filters for the two tables demo.rna and demo.rna.medhist  

```{r}

meta.cc.demo.rna %>% 
  dplyr::filter(cohort %in% c("PP", "PD", "BF")) %>% 
  dplyr::filter(case_control_other_latest!="Other") %>% 
  dplyr::pull(participant_id) %>% 
  unique() %>% 
  length()
# [1] 2880

meta.cc.demo.rna.medhist %>% 
  dplyr::filter(cohort %in% c("PP", "PD", "BF")) %>% 
  dplyr::filter(case_control_other_latest!="Other") %>% 
  dplyr::pull(participant_id) %>% 
  unique() %>% 
  length()
# [1] 2923

```

# Get usable samples from metadata  

```{r}

meta.cc.demo.rna.medhist %>% colnames(.)

```

* Some of the entries for `case_control_other_latest` are other, and as such, these samples will be filtered out like `case_control_other_latest!="Other"`  
* We want all timepoints at the moment, so no filter for `visit_month`  
* Only including PP and PD in this analysis, because the BF cohort only has one timepoint  

```{r}

samples.to.filter.for = meta.cc.demo.rna %>% 
  dplyr::filter(cohort %in% c("PP", "PD")) %>% 
  dplyr::filter(case_control_other_latest!="Other") %>% 
  dplyr::pull(sample_id.x)

```

```{r}

meta.cc.demo.rna %>% 
  dplyr::filter(cohort %in% c("PP", "PD")) %>% 
  dplyr::filter(case_control_other_latest!="Other") %>% 
  dplyr::group_by(cohort, case_control_other_latest, visit_month) %>% 
  dplyr::summarise(n=n()) %>% 
  dplyr::arrange(cohort, case_control_other_latest, visit_month) %>% 
  print(n=Inf)

```

# Filter `count.mat`  

```{r, eval=FALSE}

count.mat.filt = count.mat %>% 
  dplyr::select(all_of(samples.to.filter.for)) %>% 
  as.data.frame()

```

# Generate separate case and control tpm matrices for both cohorts  

```{r, eval=FALSE}

for(i in c("PD", "PP")){
  
  case.sampleid = meta.cc.demo.rna %>% 
    dplyr::filter(cohort == i) %>% 
    dplyr::filter(case_control_other_latest=="Case") %>% 
    dplyr::pull(sample_id.x)
  
  ctrl.sampleid = meta.cc.demo.rna %>% 
    dplyr::filter(cohort == i) %>% 
    dplyr::filter(case_control_other_latest=="Control") %>% 
    dplyr::pull(sample_id.x)
  
  cohort.sampleid = meta.cc.demo.rna %>% 
    dplyr::filter(cohort == i) %>% 
    dplyr::filter(case_control_other_latest %in% c("Control", "Case")) %>% 
    dplyr::pull(sample_id.x)
  
  # count.mat.filt %>%
  #   dplyr::select(all_of(case.sampleid)) %>%
  #   write.csv(x=., file=paste0("./data/processed_and_unprocessed_transcriptomic/",i,"_case_all_timepoint_featurecount_matrix.csv"))
  # 
  # count.mat.filt %>%
  #   dplyr::select(all_of(ctrl.sampleid)) %>%
  #   write.csv(x=., file=paste0("./data/processed_and_unprocessed_transcriptomic/",i,"_ctrl_all_timepoint_featurecount_matrix.csv"))
  
  count.mat.filt %>%
    dplyr::select(all_of(cohort.sampleid)) %>%
    write.csv(x=., file=paste0("./data/processed_and_unprocessed_transcriptomic/",i,"_cohort_all_timepoint_featurecount_matrix.csv"))
  
}

```

* Check file format before running pipeline  

```{r}

pd.case = vroom::vroom(paste0("./data/processed_and_unprocessed_transcriptomic/PD_case_all_timepoint_featurecount_matrix.csv"))

```


# Run `counts2TPM`  

* Implement python function `counts2TPM` in a bash loop to get `_TPM.csv` files  
* This code needs to be run on the CMD line, rather than within the `.rmd`  
* I used the commands in the script `~/projects/amppd_analysis/scripts/create_conda_env_mt_nuc.sh` to generate the necessary conda environment in which to run the pipeline  

```{bash, eval=F, echo=T}

# activate env containing python3.7, pandas, fire
conda activate py37 

# go to wd
cd /home/abrowne/projects/amppd_analysis/data/processed_and_unprocessed_transcriptomic  

# run counts2TPM
for i in $(ls *{case,ctrl}_all_timepoint_featurecount_matrix.csv); do echo $i; study_label=${i%%_feature*}; python3.7 /home/abrowne/projects/amppd_analysis/pipeline/transcriptomics/01-counts2TPM.py --file_dir="./" --gene_length_dir="/home/abrowne/projects/amppd_analysis/data/amppd_gene_lengths.csv" --pattern=$i --filesep="," --outdir="./mt_nuc_corr_pipeline_output/all_timepoints/" --outlabel=$study_label; done

for i in $(ls *cohort_all_timepoint_featurecount_matrix.csv); do echo $i; study_label=${i%%_feature*}; python3.7 /home/abrowne/projects/amppd_analysis/pipeline/transcriptomics/01-counts2TPM.py --file_dir="./" --gene_length_dir="/home/abrowne/projects/amppd_analysis/data/amppd_gene_lengths.csv" --pattern=$i --filesep="," --outdir="./mt_nuc_corr_pipeline_output/all_timepoints/" --outlabel=$study_label; done

```

# Check `counts2TPM` output  

```{r, eval=FALSE}

vroom("./data/processed_and_unprocessed_transcriptomic/mt_nuc_corr_pipeline_output/all_timepoints/PD_ctrl_all_timepoint_TPM.csv", show_col_types = FALSE) %>% 
  tibble::column_to_rownames("...1") %>% 
  .[1:5,1:5]

```

# Run `log10MedNormalise()`  

* This function filters out samples with 0 TPM in all genes and genes with 0 TPM in all samples  
* remove individuals that have zero reads  
`print("removing individuals with 0 reads")`  
`test = apply(df, 2, function(x) { all(x > 0)})`  
* remove genes that have zero reads  
`print("removing genes with 0 reads")`  
`test = apply(df, 1, function(x) { all(x > 0)})`    
* Generate `_log10_mediannorm_TPM.csv` and `_log10_norm_TPM.csv` files  


```{r, eval=F, echo=T}

source("./pipeline/transcriptomics/02-log10MedNormalise.R")

for(i in c("PD", "PP")){ 
  
  for(j in c("case", "ctrl")){
  # for(j in c("cohort")){
    
    log10MedNormalise(
      file_dir="./data/processed_and_unprocessed_transcriptomic/mt_nuc_corr_pipeline_output/all_timepoints/", 
      pattern=paste0(i,"_",j,"_all_timepoint_TPM.csv"), 
      filesep=",", 
      outlabel=paste0(i,"_",j,"_all_timepoint"), 
      med_norm=TRUE
    )
  }
}

```

* Remove trailing _ from filenames: e.g. `_X_Y_log10_mediannorm_TPM.csv`  
* Run this in the Terminal, not via the .rmd   

```{bash, eval=FALSE, include=TRUE}

cd /home/abrowne/projects/amppd_analysis/data/mt_nuc_corr_pipeline_output/

for f in _*; do mv -- "$f" "${f##*(_)}"; done

```

# Session Info

```{r}
sessionInfo()
```

