---
title: "AMPPD-analyse-annotated-matrixEQTL-files"
author: "Aine Fairbrother-Browne"
date: ""
output: 
  html_document:
    theme: paper
    highlight: kate
    df_print: paged
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: false  
    fig_width: 20
    fig_height: 15
    fig_caption: true
---

**Aim**: To analyse the annotated `MatrixEQTL` output  

# Setup  

```{r setup, include=FALSE}

# import libs
library(tidyverse)

# notebook settings 
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/home/abrowne/projects/amppd_analysis/")

```

## Define helper functions  

```{r}

parallel.helper = function(fn, path, pattern, cores.to.use=3){
  
  # Run a function, fn, across files matching pattern, pattern, in directory, path. 
  # This will produce whatever files fn is designed to output
  file.list = list.files(path=path, pattern=pattern, full.names=T)
  parallel::mclapply(file.list, fn, mc.cores=cores.to.use)
  
}

lapply.helper = function(fn, path, pattern){
  
  # Run a function, fn, across files matching pattern, pattern, in directory, path. 
  # This will produce whatever files fn is designed to output
  file.list = list.files(path=path, pattern=pattern, full.names=T)
  lapply(X=file.list, FUN=fn)
  
}

loop.helper = function(fn, path, pattern){
  
  file.list = list.files(path=path, pattern=pattern, full.names=T)
  for(i in 1:length(file.list)){
    fn(file.list[i])
  }
}

```

# Define function to read in and filter `MatrixEQTL` results    

* Currently each file has 24 variables, which include: basic info (cohort, diagnosis, model), MatrixEQTL output stats, rsIDs, variantIDs, nearest gene info  
* But the files are large, and contain lots of non-significant (5x10-8<P<0.05) data that can be removed  
* This will allow R to handle the data more easily  
* So read in each file, filter appropriately, and add to list of dataframes   
* Then merge into large dataframe  

```{r eval=FALSE, message=FALSE, warning=FALSE}

read.annot.file.apply.filter = function(file.path, p.cutoff=1e-03){
  
  d = vroom::vroom(file.path) 
  
  if( !("nearestGene.id" %in% colnames(d))) {
    d = d %>% 
      dplyr::mutate(nearestGene.symbol=NA, 
                    nearestGene.id=NA,
                    nearestGene.bioType=NA)
  }
  d %>%
    dplyr::filter(p.value<p.cutoff) %>%
    # make snp col a number
    dplyr::mutate(snp.chr=as.numeric(snp.chr)) %>% 
    return(.)
}

```

# Run `read.annot.file.apply.filter()`
* This collects and filters results before collating into a table  

## Run for the different correction strategy outputs  

### 1. Standard correction  

5 x gPC  
Sex  
Age  
Diagnosis (for cohort, case+control)  
10 x PEER  

```{r eval=FALSE, message=FALSE, warning=FALSE}

lapply.helper(fn=read.annot.file.apply.filter, path="./data/MatrixEQTL_output", pattern="_wrangled.csv", p.cutoff=1e-03) %>% 
  # then filter list of files and bind them together
  .[lapply(., length) > 0] %>% 
  .[map(., ~dim(.)[1]) > 0] %>% 
  dplyr::bind_rows() %>% 
  # remove extra cols 
  dplyr::select(-(ends_with(".x"))) %>% 
  dplyr::select(-(ends_with(".y"))) %>% 
  dplyr::select(-contains("nearestGene")) %>% 
  readr::write_csv(x=., file="./data/MatrixEQTL_output/aggregated_tables/matrixeqtl_res_aggregated_p1e-03_filter.csv")

```

### 2. Standard + celltype (no PEER axes)

5 x gPC  
Sex  
Age  
Diagnosis (for cohort, case+control)  
Celltype proportions  

```{r eval=FALSE, message=FALSE, warning=FALSE}

lapply.helper(fn=read.annot.file.apply.filter, path="./data/MatrixEQTL_output/celltype_plus_standard_correction_no_PEER/", pattern="_wrangled.csv") %>% 
  # then filter list of files and bind them together
  .[lapply(., length) > 0] %>% 
  .[map(., ~dim(.)[1]) > 0] %>% 
  dplyr::bind_rows() %>% 
  readr::write_csv(x=., file="./data/MatrixEQTL_output/celltype_plus_standard_correction_no_PEER/aggregated_tables/matrixeqtl_res_aggregated_GW_filter.csv")

```

### 3. Standard + celltype (with PEER axes)  

5 x gPC  
Sex  
Age  
Diagnosis (for cohort, case+control)  
10 x PEER  
Celltype proportions  

```{r eval=FALSE, message=FALSE, warning=FALSE}

lapply.helper(fn=read.annot.file.apply.filter, path="./data/MatrixEQTL_output/celltype_plus_standard_correction_with_PEER/", pattern="_wrangled.csv") %>% 
  # then filter list of files and bind them together
  .[lapply(., length) > 0] %>% 
  .[map(., ~dim(.)[1]) > 0] %>% 
  dplyr::bind_rows() %>% 
  readr::write_csv(x=., file="./data/MatrixEQTL_output/celltype_plus_standard_correction_with_PEER/aggregated_tables/matrixeqtl_res_aggregated_GW_filter.csv")

```

# Read in aggregated `MatrixEQTL` results  

* Read in files generated by `read.annot.file.apply.filter()`  
* Also read in `.bim` file corresponding to the PPMI cohort - this contains rsIDs, variant IDs and alt and ref allele information  

```{r}

pp.bim = dplyr::bind_rows(
  genio::read_bim(file="/home/abrowne/projects/amppd_analysis/data/genomics_bed_bim_fam_from_rosalind/PP_all_chrs_maf0.05.bim", verbose = TRUE)) %>% 
  # see plink REF/ALT behaviour - https://www.biostars.org/p/9495739/
  dplyr::mutate(variant.id=paste0(chr, "_", pos, "_", alt, "_", ref)) %>% 
  dplyr::rename(snp.chr=chr, snp.start=pos)

pp.bim %>% head()

```

* Read in results generated with correction strategy 1 (original correction strategy)   

```{r message=FALSE, warning=FALSE}

eqtl.res.filtered = readr::read_csv(file="~/projects/amppd_analysis/data/MatrixEQTL_output/aggregated_tables/matrixeqtl_res_aggregated_p1e-03_filter.csv") %>%
  dplyr::mutate(snp.chr=as.character(snp.chr)) %>% 
  dplyr::filter(cohort=="PP") %>% 
  # fix the X, Y chromosome NA problem 
  tidyr::separate(col=snp.chr.pos, into=c("snp.chr", "snp.pos"), sep=":", remove=F) %>% 
  dplyr::mutate(snp.chr=gsub("chr", "", snp.chr)) %>% 
  # join the bim information  
  dplyr::left_join(x=., y=pp.bim, by=c("snp.chr", "snp.start"))

eqtl.res.filtered %>% 
  head()

```

# Explore strategy 1 results  

* Get the numbers of genome wide sig. hits for each `diagnosis`-`phenotype.category` group   

```{r message=FALSE, warning=FALSE}

eqtl.res.filtered %>% 
  dplyr::filter(p.value<5e-08) %>% 
  dplyr::group_by(cohort, diagnosis, phenotype.category) %>% 
  dplyr::summarise(n=n())

```

* Could draw a cut-off for 'trustworthy' MT sites at min. 25% non-missing data in case and control  
* These sites are 5647, 5721, 7526, 9999, 5883, 5818, 2617  
* Filtering for these  

```{r}

eqtl.res.filtered = eqtl.res.filtered %>% 
  # when the phenotype.category is methylation, filter for particular sites
  dplyr::filter( (phenotype.category == "methylation") & (phenotype %in% paste0(c(5647, 5721, 7526, 9999, 5883, 5818, 2617), "_full")) | (phenotype.category == "expression") )

```

* Get number of hits for each `diagnosis`-`phenotype.category` group post-filtration  

```{r message=FALSE, warning=FALSE}

eqtl.res.filtered %>% 
  dplyr::group_by(cohort, diagnosis, phenotype.category) %>% 
  dplyr::summarise(n=n())

```

* Look at top 10 hits  

```{r}

eqtl.res.filtered %>% 
  dplyr::arrange(FDR) %>% 
  dplyr::select(snp.chr.pos, phenotype, beta, p.value, FDR, cohort, diagnosis, SNP, variant.id) %>% 
  head(n=10)

```

# Annotate top hits with opentargets information  

* Import functions  
* `query.opentargets.with.varIDs()`  
* `query.opentargets.with.varIDs.helper()`  

```{r}

source("/home/abrowne/projects/amppd_analysis/scripts/query-api-genetics-opentargets-with-helper-funs.R")

```

* Run function `query.opentargets.with.varIDs()` which employs `query.opentargets.with.varIDs.helper()`  
* This annotates the results with nearest gene ID, symbol and biotype  

```{r}

eqtl.res.filtered.annotated = lapply(X=eqtl.res.filtered$variant.id, FUN=query.opentargets.with.varIDs) %>%
  dplyr::bind_rows() %>%
  tibble::as_tibble() %>%
  dplyr::mutate(snp.chr.pos=paste0("chr",chromosome,":",position)) %>%
  dplyr::select(any_of( c("snp.chr.pos", "nearestGene.id", "nearestGene.symbol", "nearestGene.bioType") )) %>%
  dplyr::distinct() %>%
  dplyr::left_join(x=eqtl.res.filtered , y=., by="snp.chr.pos") %>% 
  dplyr::select(-(ends_with(".x"))) %>% 
  dplyr::rename_all(~str_replace_all(.,"\\.y",""))

```

* Look at methylation results, sorted by `FDR`  

```{r}

eqtl.res.filtered %>% 
  dplyr::filter(phenotype.category=="methylation") %>% 
  dplyr::arrange(p.value) %>% 
  dplyr::select(SNP, phenotype, variant.id, chr.mb.block.id, diagnosis, beta, p.value, p.value.meta, FDR, contains("nearest")) %>% 
  print(n=Inf)

```

* Peak SNP in mb block only  

```{r}

eqtl.res.filtered %>% 
  dplyr::filter(phenotype.category=="methylation") %>% 
  dplyr::group_by(cohort, diagnosis, phenotype, chr.mb.block.id) %>% 
  dplyr::arrange(p.value) %>% 
  dplyr::slice_head() %>% 
  dplyr::ungroup() %>% 
  dplyr::select(SNP, phenotype, variant.id, chr.mb.block.id, diagnosis, beta, p.value, p.value.meta, FDR, contains("nearest")) %>% 
  dplyr::arrange(FDR)

```

* Get numbers of hits when reduced to representative peak SNP  

```{r}

eqtl.res.filtered %>% 
  dplyr::filter(phenotype.category=="methylation") %>% 
  dplyr::group_by(cohort, diagnosis, phenotype, chr.mb.block.id) %>% 
  dplyr::arrange(p.value) %>% 
  dplyr::slice_head() %>% 
  dplyr::ungroup() %>% 
  dplyr::group_by(cohort, diagnosis, phenotype.category) %>% 
  dplyr::summarise(n=n())

```

* Look at expression results, sorted by `FDR`  

```{r}

eqtl.res.filtered %>% 
  dplyr::filter(phenotype.category=="expression") %>% 
  dplyr::arrange(p.value) %>% 
  dplyr::select(SNP, phenotype, variant.id, chr.mb.block.id, diagnosis, beta, p.value, p.value.meta, FDR, contains("nearest")) %>% 
  print(n=Inf)

```

* Peak SNP in mb block only  

```{r}

eqtl.res.filtered %>% 
  dplyr::filter(phenotype.category=="expression") %>% 
  dplyr::arrange(p.value) %>% 
  dplyr::group_by(cohort, diagnosis, phenotype, chr.mb.block.id) %>% 
  dplyr::slice_head() %>% 
  dplyr::select(SNP, phenotype, variant.id, diagnosis, beta, p.value, p.value.meta, FDR, contains("nearest")) %>% 
  dplyr::arrange(FDR)

```

* Get numbers of hits when reduced to representative peak SNP  

```{r}

eqtl.res.filtered %>% 
  dplyr::filter(phenotype.category=="expression") %>% 
  dplyr::group_by(cohort, diagnosis, phenotype, chr.mb.block.id) %>% 
  dplyr::arrange(p.value) %>% 
  dplyr::slice_head() %>% 
  dplyr::ungroup() %>% 
  dplyr::group_by(cohort, diagnosis, phenotype.category) %>% 
  dplyr::summarise(n=n())

```

# Visualise top hits  

## Define `plot.indiv.assocs`

* This is a function to pull out raw data for selected SNP-gene or SNP-MTpos combination, run lm() and plot

```{r, eval=F}

plot.indiv.assocs = function(pheno, snp_chr_pos, exp_or_meth, cohort, diagnosis){
  
  # pheno: ENS code for the MT gene from which expression is derived, or MT position for which the methylation prop. is derived
  # snp_chr_pos: chr:pos format for SNP location
  # exp_or_meth: looking at expression or methylation data
  # cohort:
  # diagnosis:
  
  snp_chr_pos = gsub(":", "-", snp_chr_pos)
  
  if(diagnosis=="cohort"){
    s1 = vroom::vroom(paste0("/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/genomics/celltype_PP_samples_filtered/", cohort, "_cohort_all_chrs-maf0.05_geno_matrix_012_wrangled-8b.csv")) %>%
      tibble::column_to_rownames("pos") %>%
      .[snp_chr_pos, ]
  } else{
    s1 = vroom::vroom(paste0("/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/genomics/celltype_PP_samples_filtered/", cohort, "_", tolower(diagnosis), "_all_chrs-maf0.05_geno_matrix_012_wrangled-8b.csv")) %>%
      tibble::column_to_rownames("pos") %>%
      .[snp_chr_pos, ]
  }
  if(exp_or_meth=="expression"){
    e1 = vroom::vroom(paste0("/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/transcriptomics/celltype_PP_samples_filtered/", cohort, "_", diagnosis, "_pheno=",pheno,"_log10_mediannorm_TPM_masked_outliers_q1-3iqr_q3+3iqr_MT_ONLY.csv")) %>%
      dplyr::filter(if_any(everything(), ~ !is.na(.))) %>%
      tibble::column_to_rownames("sample_id") %>%
      `colnames<-`(colnames(s1))
  }
  if(exp_or_meth=="methylation"){
    e1 = vroom::vroom(paste0("/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/methylomics/celltype_PP_samples_filtered/", cohort, "_", diagnosis, "_pheno=", pheno,"_methylation_matrix_MatrixEQTL_input.csv")) %>%
      dplyr::filter(if_any(everything(), ~ !is.na(.))) %>%
      tibble::column_to_rownames("sample_id") %>%
      `colnames<-`(colnames(s1))
  }
  
  cov = vroom::vroom(paste0("/home/abrowne/projects/amppd_analysis/data/MatrixEQTL_input/covariates/celltype_PP_samples_filtered/", cohort, "_", diagnosis, "_maf0.05_cov_table_for_MatrixEQTL.csv")) %>%
    tibble::column_to_rownames("variable") %>%
    `colnames<-`(colnames(s1))
  
  all = rbind(e1, s1, cov) %>%
    t() %>%
    as.data.frame()
  
  formula = all[,pheno] ~ all[,snp_chr_pos] + all[, "gPC1"] + all[, "gPC2"] + all[, "gPC3"] + all[, "gPC4"] + all[, "gPC5"] + all[, "PEER1"] + all[, "PEER2"] + all[, "PEER3"] + all[, "PEER4"] + all[, "PEER5"] + all[, "PEER6"] + all[, "PEER7"] + all[, "PEER8"] + all[, "PEER9"] + all[, "PEER10"] + all[, "sex"] + all[, "age_at_baseline"]
  
  lm1 = lm(formula=formula,
           data=all,
           na.action="na.exclude")
  
  # to allow vars to be used to select cols within ggplot call
  snp_chr_pos = sym(snp_chr_pos)
  pheno=sym(pheno)
  
  return(
    all %>% 
      tidyr::drop_na() %>% 
      dplyr::mutate(geno.label=case_when(
        snp_chr_pos==0 ~ "AA",
        snp_chr_pos==1 ~ "AB",
        snp_chr_pos==2 ~ "BB"
      )) %>% 
      tibble::rownames_to_column("sample.id") %>% 
      dplyr::filter(sample.id %in% names(lm1$fitted.values)) %>% 
      
      ggplot(data=., aes(x=factor(!!snp_chr_pos), y=!!pheno)) +
      theme_minimal() +
      geom_boxplot() +
      geom_point(position="jitter", alpha=0.5) +
      geom_smooth(mapping=aes(group=1), method="lm", se=T, fullrange=F) +
      xlab("Genotype") +
      ylab(pheno)
  )
  
}
```

# Are hits maintained across correction strategies?  

## Load strategy 2 results  

```{r}

celltype.no.peer = readr::read_csv(file="~/projects/amppd_analysis/data/MatrixEQTL_output/celltype_plus_standard_correction_no_PEER/aggregated_tables/matrixeqtl_res_aggregated_GW_filter.csv") %>% 
  dplyr::mutate(snp.chr=as.character(snp.chr)) %>% 
  dplyr::filter(cohort=="PP") %>% 
  # fix the X, Y chromosome NA problem 
  tidyr::separate(col=snp.chr.pos, into=c("snp.chr", "snp.pos"), sep=":", remove=F) %>% 
  dplyr::mutate(snp.chr=gsub("chr", "", snp.chr)) %>% 
  # join the bim information  
  dplyr::left_join(x=., y=pp.bim, by=c("snp.chr", "snp.start")) %>% 
  dplyr::mutate(unique.association.id = paste0(cohort,"-",diagnosis,"-",phenotype,"-",snp.chr,"-",snp.start))


```

## Load strategy 3 results  

```{r}

celltype.with.peer = readr::read_csv(file="~/projects/amppd_analysis/data/MatrixEQTL_output/celltype_plus_standard_correction_with_PEER/aggregated_tables/matrixeqtl_res_aggregated_GW_filter.csv") %>% 
  dplyr::mutate(snp.chr=as.character(snp.chr)) %>% 
  dplyr::filter(cohort=="PP") %>% 
  # fix the X, Y chromosome NA problem 
  tidyr::separate(col=snp.chr.pos, into=c("snp.chr", "snp.pos"), sep=":", remove=F) %>% 
  dplyr::mutate(snp.chr=gsub("chr", "", snp.chr)) %>% 
  # join the bim information  
  dplyr::left_join(x=., y=pp.bim, by=c("snp.chr", "snp.start")) %>% 
  dplyr::mutate(unique.association.id = paste0(cohort,"-",diagnosis,"-",phenotype,"-",snp.chr,"-",snp.start))

```

## Visualise the comparison    

```{r}

# add unique.association.id to allow cross-strategy comparison
eqtl.res.filtered = eqtl.res.filtered %>% 
  dplyr::mutate(unique.association.id = paste0(cohort,"-",diagnosis,"-",phenotype,"-",snp.chr,"-",snp.start))

```

* Get set of `unique.association.id` present in all 3 analyses  
* These are associations that passed p<1e-03 in all 3 analyses  

```{r}

set.of.assocs = Reduce(intersect, list(celltype.no.peer %>% dplyr::pull(unique.association.id),
                                       celltype.with.peer %>% dplyr::pull(unique.association.id),
                                       eqtl.res.filtered %>% dplyr::pull(unique.association.id)
))

celltype.no.peer %>% dplyr::pull(unique.association.id) %>% length()
celltype.with.peer %>% dplyr::pull(unique.association.id) %>% length()
eqtl.res.filtered %>% dplyr::pull(unique.association.id) %>% length()

set.of.assocs %>% length()

```

* Around 40% of associations are present in all 3 analyses at p<1e-03  
* Make new df of p.values  

```{r}

p.val.comparison.df = data.frame(
  
  celltype.no.peer = celltype.no.peer %>% 
    dplyr::filter(unique.association.id %in% set.of.assocs) %>% 
    dplyr::arrange(match(unique.association.id, set.of.assocs)) %>% 
    dplyr::pull(p.value),
  
  celltype.with.peer = celltype.with.peer %>% 
    dplyr::filter(unique.association.id %in% set.of.assocs) %>%
    dplyr::arrange(match(unique.association.id, set.of.assocs)) %>% 
    dplyr::pull(p.value),
  
  eqtl.res.filtered = eqtl.res.filtered %>% 
    dplyr::filter(unique.association.id %in% set.of.assocs) %>% 
    dplyr::arrange(match(unique.association.id, set.of.assocs)) %>% 
    dplyr::pull(p.value)
  
)

p.val.comparison.df %>% head()

```

```{r fig.height=10, fig.width=6}

library(patchwork)

p1 = p.val.comparison.df %>% 
  ggplot(data=., aes(x=eqtl.res.filtered, y=celltype.with.peer)) + 
  geom_point(alpha=0.005) +
  xlab("standard correction") +
  ylab("standard correction + celltype")
p2 = p.val.comparison.df %>% 
  ggplot(data=., aes(x=eqtl.res.filtered, y=celltype.no.peer)) + 
  geom_point(alpha=0.005) +
  xlab("standard correction") +
  ylab("standard correction + celltype, - peer")
p3 = p.val.comparison.df %>% 
  ggplot(data=., aes(x=celltype.with.peer, y=celltype.no.peer)) + 
  geom_point(alpha=0.005) +
  xlab("standard correction + celltype") + 
  ylab("standard correction + celltype, - peer")

p1 / p2 / p3

```

```{r fig.width=10}

comp1 = p.val.comparison.df %>% 
  tidyr::pivot_longer(cols=everything(), names_to = "label", values_to="p.value") %>% 
  ggplot(data=., aes(x=p.value, fill=label)) + 
  geom_density(alpha=0.5) +
  geom_vline(xintercept=5e-08) + 
  xlim(0, 5e-07) + 
  theme(legend.position = "none")

comp2 = p.val.comparison.df %>% 
  tidyr::pivot_longer(cols=everything(), names_to = "label", values_to="p.value") %>% 
  ggplot(data=., aes(x=p.value, fill=label)) + 
  geom_density(alpha=0.5)

comp1 | comp2

```

* The correction strategies seem to make little difference to the distribution of p-values (for the set of associations passing p<5e-04 in each correction strategy res, and that are common between the 3 strategies)  
* The celltype+peer and standard (batch+peer) strategies appear similar  
* I will use the standard correction as the discovery set, cross-referencing top hits with these to ensure they are not driven by celltype  





